asciidoc/session-API.adoc

```diff
@@ -0,0 +1,682 @@
+----
+======
+
+[.include-with-javascript]
+======
+[source, javascript]
+----
+include::{javascript-examples}/examples.test.js[tags=result-consume]
+----
+======
+
+// [.include-with-python]
+// ======
+// [source, python]
+// ----
+// include::{python-examples}/result_consume_example.py[tags=result-consume]
+// ----
+// ======
+====
+
+
+[[driver-simple-result-retain]]
+=== Retaining results
+
+Within a session, only one result stream can be active at any one time.
+Therefore, if the result of one query is not fully consumed before another query is executed, the remainder of the first result will be automatically buffered within the Result object.
+
+image::driver-result-buffer.svg[title="Result buffer", role="middle"]
+
+This buffer provides a staging point for results, and divides result handling into _fetching_ (moving from the network to the buffer) and _consuming_ (moving from the buffer to the application).
+
+[NOTE]
+For large results, the result buffer may require a significant amount of memory. For this reason, it is recommended to consume results in order wherever possible.
+
+Client applications can choose to take control of more advanced query patterns by explicitly retaining results.
+Such explicit retention may also be useful when a result needs to be saved for future processing.
+The drivers offer support for this process, as per the example below:
+
+[.tabbed-example]
+.Retain results for further processing
+====
+[.include-with-dotnet]
+======
+[source, csharp]
+----
+include::{dotnet-examples}/Examples.cs[tags=result-retain]
+----
+======
+// [.include-with-go]
+// ======
+// [source, go]
+// ----
+// include::{go-examples}/examples_test.go[tags=result-retain]
+// ----
+// ======
+[.include-with-java]
+======
+[source, java]
+----
+include::{java-examples}/ResultRetainExample.java[tags=result-retain]
+----
+======
+[.include-with-javascript]
+======
+[source, javascript]
+----
+include::{javascript-examples}/examples.test.js[tags=result-retain]
+----
+======
+// [.include-with-python]
+// ======
+// [source, python]
+// ----
+// include::{python-examples}/result_retain_example.py[tags=result-retain]
+// ----
+// ======
+====
+
+
+[[driver-async-sessions]]
+== Asynchronous sessions
+
+
+[NOTE]
+Available in .NET, Java and JavaScript.
+
+Asynchronous sessions provide an API wherein function calls typically return available objects such as futures.
+This allows client applications to work within asynchronous frameworks and take advantage of cooperative multitasking.
+
+[[driver-async-lifecycle]]
+=== Lifecycle
+
+Session lifetime begins with session construction.
+A session then exists until it is closed, which is typically set to occur after its contained query results have been consumed.
+
+Sessions can be configured in a number of different ways. This is carried out by supplying configuration inside the session constructor.
+See 4.4. Session configuration for more details.
```

Comment:
See <<driver-session-configuration>> for more details.

Reply:
yep